packages/datadog-plugin-couchbase/src/index.js

```diff
@@ -0,0 +1,312 @@
+    }
+  }
+}
+
+function createWrapN1qlRequest (tracer) {
+  return function wrapN1qlRequest (_n1qlReq) {
+    return function n1qlRequestWithTrace (host) {
+      const span = tracer.scope().active()
+
+      addBucketTag(this, span)
+      span.setTag('cluster.host', host)
+
+      return _n1qlReq.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapViewQuery (tracer, config) {
+  return function wrapViewQuery (_view) {
+    return function viewQueryWithTrace () {
+      const viewName = arguments[2]
+      const callback = arguments[_view.length - 1]
+      const scope = tracer.scope()
+      const span = startQuerySpan(tracer, config, 'view', viewName)
+
+      if (callback) {
+        arguments[_view.length - 1] = tx.wrap(span, callback)
+      }
+
+      const req = scope.bind(_view, span).apply(this, arguments)
+
+      onRequestFinish(req, span)
+
+      return scope.bind(req)
+    }
+  }
+}
+
+function createWrapViewRequest (tracer) {
+  return function wrapViewRequest (_viewReq) {
+    return function viewRequestWithTrace () {
+      const span = tracer.scope().active()
+      const ddoc = arguments[1]
+
+      addBucketTag(this, span)
+      span.setTag('ddoc', ddoc)
+
+      return _viewReq.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapFtsQuery (tracer, config) {
+  return function wrapFtsQuery (_fts) {
+    return function ftsQueryWithTrace (query) {
+      const callback = arguments[_fts.length - 1]
+      const scope = tracer.scope()
+      const index = query.data.indexName
+      const span = startQuerySpan(tracer, config, 'search', index)
+
+      if (callback) {
+        arguments[_fts.length - 1] = tx.wrap(span, callback)
+      }
+
+      const req = scope.bind(_fts, span).apply(this, arguments)
+
+      onRequestFinish(req, span)
+
+      return scope.bind(req)
+    }
+  }
+}
+
+function createWrapFtsRequest (tracer) {
+  return function wrapFtsRequest (_ftsReq) {
+    return function ftsRequestWithTrace () {
+      const span = tracer.scope().active()
+
+      addBucketTag(this, span)
+
+      return _ftsReq.apply(this, arguments)
+    }
+  }
+}
+
+function createWrapCbasQuery (tracer, config) {
+  return function wrapCbasQuery (_cbas) {
+    return function cbasQueryWithTrace (query) {
+      const callback = arguments[_cbas.length - 1]
+      const scope = tracer.scope()
+      const cbasQuery = query.options.statement
+      const span = startQuerySpan(tracer, config, 'cbas', cbasQuery)
+
+      if (callback) {
+        arguments[_cbas.length] = tx.wrap(span, callback)
+      }
+
+      const req = scope.bind(_cbas, span).apply(this, arguments)
+
+      onRequestFinish(req, span)
+
+      return scope.bind(req)
+    }
+  }
+}
+
+function createWrapCbasRequest (tracer) {
+  return function wrapCbasRequest (_cbasReq) {
+    return function cbasRequestWithTrace (host) {
+      const span = tracer.scope().active()
+
+      addBucketTag(this, span)
+      span.setTag('cbas.host', host)
```

Comment:
We can get the port, but unfortunately, we can't get the `host` for other queries.

You can supply a host or a list of hosts to Couchbase JS and it will in turn request the cluster topology. I'm not 100% sure but it looks like it can pick any one of them to send the request to. It does all of this in native code.

Reply:
Following up on this, the `host` parameter is almost always going to be `undefined` so we actually can't get the `out.host` or `out.port` for any query except for `cbas` queries.